### PR TITLE
feat(vis-type: bar): download chart is clipped for long charts

### LINKED ISSUE #479: Download chart is clipped for long charts
### Environment

 * Release number or git hash: 
 * Browser: 
 * Deployed / Local: 
---

### Steps to reproduce the bug

1. Open a bar chart that has scrollbars
2. Download the image

### Observed Behavior

The downloaded image is clipped.

Bar chart (4)

### Expected Behavior

The downloaded image should not be clipped.

### PR DESCRIPTION
**Pull request by dv-usama-ansari**
Closes #479

### Developer Checklist (Definition of Done)

**Issue**

- [x] All acceptance criteria from the issue are met
- [x] Tested in latest Chrome/Firefox

**UI/UX/Vis**

- [x] Requires UI/UX/Vis review
  - [x] Reviewer(s) are notified @thinkh @dvchristianbors 
  - [ ] Review has occurred (_link to notes_)
  - [ ] Feedback is included in this PR
  - [ ] Reviewer(s) approve of concept and design

**Code**

- [x] Branch is up-to-date with the branch to be merged with, i.e., develop
- [x] Code is cleaned up and formatted
- [ ] Unit tests are written (frontend/backend if applicable)
- [ ] Integration tests are written (if applicable)

**PR**

- [x] Descriptive title for this pull request is provided (will be used for release notes later)
- [x] Reviewer and assignees are defined
- [x] Add type label (e.g., *bug*, *feature*) to this pull request
- [x] Add release label (e.g., `release: minor`) to this PR following [semver](https://semver.org/)
- [x] The PR is connected to the corresponding issue (via `Closes #...`)
- [x] [Summary of changes](#summary-of-changes) is written

### Summary of changes

- Added download feature for bar chart. Please see screenshots section for detailed implementation.

### Screenshots
- Long screenshot: 
  - [Screencast](https://drive.google.com/file/d/10meuGMDqY_AKjmZYllNpmjAj0anCquzh/view?usp=drive_link)
  - [Downloaded file](https://drive.google.com/file/d/1GKUR5valH6pRhHbYOCT4xfBsbrnfVJKB/view?usp=drive_link)

- Focused facet:
  - [Screencast](https://drive.google.com/file/d/1W1CMpCtVxMR_NACAitWaWdmgv06JrkZp/view?usp=drive_link)
  - Downloaded files: 
    - https://drive.google.com/file/d/1y0l1E_8inWEEnxQewyiu4s1fMeRfjFSj/view?usp=drive_link
    - https://drive.google.com/file/d/1hnYR5StVWDPvdVQ1nKvYTMYT-K2y6Mbp/view?usp=drive_link

- Small Facets fitting in the screen:
  - [Screencast](https://drive.google.com/file/d/1Lxtwb_IKwxCzPW-1oLYuE7k-6nKryBFl/view?usp=drive_link)
  - [Downloaded file](https://drive.google.com/file/d/1u8Im0TWaJ5qgNhVVWK3_jyOObFgpaAZ1/view?usp=drive_link)

- Many facets having less bars:
  - [Screencast](https://drive.google.com/file/d/16t1r8Nsvb2P-5P2zQcQffmxga6GLkK4D/view?usp=drive_link)
  - [Downloaded file](https://drive.google.com/file/d/1-9bK7y_1_zXWY0m9zGarASE-A89skFuD/view?usp=drive_link)

- Many facets having a lot of bars:
  - [Screencast](https://drive.google.com/file/d/1_l-jDRB0iGFaczMfZh9RoPZU5Jch8nKp/view?usp=drive_link)
  - [Downloaded file](https://drive.google.com/file/d/19-VPSrEHTK1O1RyZjQwIryD-oSePmj5C/view?usp=drive_link)

### Additional notes for the reviewer(s)

- Pasting links to google drive since the size of individual screencasts are very large and GitHub does not support uploading files greater than 10MB to the description.
- Using library [JSZip](https://www.npmjs.com/package/jszip) for zipping the files before downloading them.
- JSZip is large to import `(~123.8kB, 36.1kB gzipped)`, we should load it lazily: what do you think @puehringer @thinkh?

Screenshot from 2024-10-09 07-55-15

### CODE REVIEW

File: src/vis/useCaptureVisScreenshot.ts [hunk 1]
```diff
@@ -17,25 +23,81 @@ export function useCaptureVisScreenshot(uniquePlotId: string, visConfig: BaseVis
...
+                canvasWidth: canvas.width,
+                canvasHeight: canvas.height,
+                cacheBust: true,
+              });
+              return blob;
+            }),
+          );
+          blobList.forEach((blob, i) => {
+            if (blob) {
+              const fileName = `${config?.facets?.name}: ${boxList[i]?.dataset?.facet} | ${config?.aggregateType === EAggregateTypes.COUNT ? config?.aggregateType : `${config?.aggregateType} of ${config?.aggregateColumn?.name}`}: ${config?.catColumnSelected?.name}`;
```
Comment (thinkh) [inline src/vis/useCaptureVisScreenshot.ts:69]: Taking the column or facet name is generally a good idea and makes the lookup in a zip file easier for users. However, these names might contain special characters and even problematic characters for OS (e.g., `/`, `.`, `:`, `|`, ...). We need to sanitize the file name, e.g., by using this library: https://github.com/parshap/node-sanitize-filename/blob/master/index.js

**Review by thinkh — CHANGES_REQUESTED** [review]
@dv-usama-ansari Thanks for adding the download feature, which looks and works very nice. I especially like the zip feature.

As discussed in the standup today, we can keep the functionality like this for now and improve it in the next iteration. From the user's perspective screenshots of _all_ facets must be included in the zip file (and not only the ones in the view port). Please take a look at or after your innovation day about offscreen canvas and follow-up on this.

We discussed in the core meeting that you don't need to lazy load the jszip library, so your current import is fine.

However, I'd like to see some PW tests for the download functionality. There should be already some tests for the scatterplot. You could copy and adapt the download test that @oltionchampari improved recently in Ordino:

https://github.com/datavisyn/reprovisyn/blob/3a6d903f044ddbb45e3f53f32222557f30343602/playwright/05-workbenches/07-download-plot.spec.ts#L13-L19

File: src/vis/useCaptureVisScreenshot.ts (hunk 1 above)
Comment (dv-usama-ansari) [reply src/vis/useCaptureVisScreenshot.ts]: Filename sanitization implemented in https://github.com/datavisyn/visyn_core/pull/561/commits/c39daf47352c12809e565886267f3e9db5073edd

**Review by thinkh — APPROVED** [review]
Thanks for the changes. It looks good to me know. 👍 

@dvmartinweigl please approve the PR if it works for you, too.

**Comment by dvmartinweigl** [comment]
For some reason I get empty archives in bioinsight with facts enabled.
Maybe due to the `useResponsiveBarWidth` flag?

image

**Comment by dv-usama-ansari** [comment]
> For some reason I get empty archives in bioinsight with facts enabled. Maybe due to the `useResponsiveBarWidth` flag? What do you think @dv-usama-ansari ?

This is unusual. Could we connect today to take a look at this?

**Comment by dv-usama-ansari** [comment]
> For some reason I get empty archives in bioinsight with facts enabled. Maybe due to the `useResponsiveBarWidth` flag? What do you think @dv-usama-ansari?

 Resolved in https://github.com/datavisyn/visyn_core/pull/561/commits/81c192fabdeae807dcdf0ee0d8d627cba547d656

**Comment by thinkh** [comment]
@dv-usama-ansari Thanks for the fix.

@dvmartinweigl Could you please do another round of testing? Please approve this PR if it works for you, so that we can merge it afterwards. Thanks.

**Review by thinkh — CHANGES_REQUESTED** [review]
I reject my previous approval after testing it again. I can confirm @dvmartinweigl's observation of empty zip files even after committing the fix. 

@dv-usama-ansari Please try to reproduce the error with the following configuration:

image

**Comment by thinkh** [comment]
I noticed that the size of the zip file changes depending on the number of facets rendered. Both files are empty when opening them with the Windows Explorer.

### Firstexample

image

[Bar chart.zip](https://github.com/user-attachments/files/17368554/Bar.chart.zip)

Opening the zip file with 7-Zip works:

image


### Second example

image

Downloads [Bar chart (1).zip](https://github.com/user-attachments/files/17368540/Bar.chart.1.zip)

Opening the zip file with 7-Zip works:

image

Extraction works with 7-Zip as well:

image

I noticed that 7-Zip automatically replaced the `:`, `|` characters on extraction. My assumption is that the file names don't work with the explorer internal zip program. 

@dv-usama-ansari Please check and revise the filename sanitation. If possible write/extend jest tests for it.

**Review by thinkh — APPROVED** [review]
Thanks for fixing the filenames in the zip file. It works now as expected and I can see and open the PNG files directly from the Windows Explorer.

image

**Pull request merged**